apps/meeting/src/providers/VideoGridProvider/index.tsx

```diff
@@ -0,0 +1,212 @@
+// Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import {
+  useActiveSpeakersState,
+  useAudioVideo,
+  useContentShareState,
+  useLocalVideo,
+  useMeetingManager,
+  useRosterState,
+} from 'amazon-chime-sdk-component-library-react';
+import {
+  AudioVideoObserver,
+  VideoDownlinkObserver,
+  VideoSource,
+} from 'amazon-chime-sdk-js';
+import React, { createContext, useContext, useEffect, useReducer } from 'react';
+import { priorityBasedPolicy } from '../../meetingConfig';
+import { VideoGridMode } from '../../types';
+import { useAppState } from '../AppStateProvider';
+import {
+  Controls,
+  initialState,
+  reducer,
+  State,
+  VideoGridAction,
+} from './state';
+
+const VideoGridStateContext = createContext<State | null>(null);
+const VideoGridControlsContext = createContext<Controls | null>(null);
+
+const VideoGridProvider: React.FC = ({ children }) => {
+  const audioVideo = useAudioVideo();
+  const meetingManager = useMeetingManager();
+  const [state, dispatch] = useReducer(reducer, initialState);
+  const { roster } = useRosterState();
+  const { videoGridMode, setVideoGridMode } = useAppState();
+  const { isVideoEnabled } = useLocalVideo();
+  const { isLocalUserSharing, sharingAttendeeId } = useContentShareState();
+  const activeSpeakers = useActiveSpeakersState();
+
+  useEffect(() => {
+    if (!audioVideo) {
+      return;
+    }
+
+    dispatch({
+      type: VideoGridAction.UpdateAttendeeStates,
+      payload: { roster },
+    });
+  }, [audioVideo, roster]);
+
+  useEffect(() => {
```

Comment:
Suggest moving this `useEffect` to the top. React executes hooks in the order they appear in an component.